qutebrowser/misc/overlap.py

```diff
@@ -0,0 +1,127 @@
+    If widget1 intersects widget2 then widget2 must intersect at least one
+    cell that widget1 intersects; this allows us to avoid checking every pair
+    of widgets to see if they intersect.
+
+    alternatives: R-tree, quadtree
+    """
+
+    def __init__(self, widgets: Iterable[QWidget]):
+        self._lookup: Dict[
+            Tuple[int, int], Set[QWidget]
+        ] = collections.defaultdict(set)
+        self._max_width = max(widget.width() for widget in widgets)
+        self._max_height = max(widget.height() for widget in widgets)
+        for widget in widgets:
+            self.register(widget)
+
+    def _cell_index(self, p: QPoint) -> Tuple[int, int]:
+        return (
+            math.floor(p.x() / self._max_width),
+            math.floor(p.y() / self._max_height),
+        )
+
+    def register(self, widget: QWidget) -> None:
+        """Add a widget."""
+        for corner in _corners(widget.geometry()):
+            self._lookup[self._cell_index(corner)].add(widget)
+
+    def deregister(self, widget: QWidget) -> None:
+        """Remove a widget."""
+        for corner in _corners(widget.geometry()):
+            self._lookup[self._cell_index(corner)].difference_update([widget])
+
+    def overlappings(self, widget: QWidget) -> Iterable[QWidget]:
+        """Get widgets that overlap with `widget`."""
+        cands = set()
+        for corner in _corners(widget.geometry()):
+            cands.update(self._lookup[self._cell_index(corner)])
+        cands.difference_update([widget])
+        return [
+            cand for cand in cands
+            if cand.geometry().intersects(widget.geometry())
+        ]
+
+
+def cycle(widgets: Sequence[QWidget], reverse: bool = False) -> None:
+    """Rotate the order of overlapping widgets.
+
+    Args:
+        widgets: List of widgets to consider. Must be siblings.
+        reverse: Cycle in reverse direction.
+    """
+    if not widgets:
+        return
+
+    children = widgets[0].parent().children()
+    widget_set = {widget for widget in widgets if widget.isVisible()}
+    if not widget_set:
+        return
+    overlap_lookup = OverlapLookup(widget_set)
+    # widget1 is above widget2 iff widget1 comes later in parent().children()
+    widgets_in_order = [child for child in children if child in widget_set]
+    if reverse:
+        widgets_in_order = list(reversed(widgets_in_order))
+
+    to_move: Set[QWidget] = set()
+    excluded: Set[QWidget] = set()
+    for widget in widgets_in_order:
+        if widget in excluded:
+            continue
+        assert isinstance(widget, QWidget)
+        overlappings = overlap_lookup.overlappings(widget)
+        if overlappings:
+            to_move.add(widget)
+            excluded.update(overlappings)
+
+    if reverse:
+        lowest = next(
+            child for child in children
+            if child in widget_set and child not in to_move
+        )
```

Comment:
This might be cleaner using a set operation?

Reply:
I don't know, any ideas?